drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt

```diff
@@ -10,84 +10,184 @@ package com.adyen.checkout.dropin.ui
 
-    private val paymentMethodsModel = PaymentMethodsModel()
+    private val nullNameFallback = "Unknown"
 
-    override fun onAvailabilityResult(isAvailable: Boolean, paymentMethod: PaymentMethod, config: Configuration?) {
-        Logger.d(TAG, "onAvailabilityResult - ${paymentMethod.type} $isAvailable")
+    private var availabilitySum = 0
+    private var availabilitySkipSum = 0
+    private var availabilityChecksum = 0
+
+    private val storedPaymentMethodsList = mutableListOf<StoredPaymentMethodModel>()
+    private val paymentMethodsList = mutableListOf<PaymentMethodModel>()
 
-        if (isAvailable) {
-            addPaymentMethod(paymentMethod)
+    fun getStoredPaymentMethod(id: String): StoredPaymentMethod {
+        for (storedPaymentMethod in paymentMethodsApiResponse.storedPaymentMethods.orEmpty()) {
+            if (storedPaymentMethod.id == id) {
+                return storedPaymentMethod
+            }
         }
+        return StoredPaymentMethod()
+    }
 
-        // TODO handle unavailable and only notify when all list is checked
+    fun getPaymentMethod(type: String): PaymentMethod {
+        for (paymentMethod in paymentMethodsApiResponse.paymentMethods.orEmpty()) {
+            if (paymentMethod.type == type) {
+                return paymentMethod
+            }
+        }
+        return PaymentMethod()
     }
 
-    private fun onPaymentMethodsResponseChanged(paymentMethods: List<PaymentMethod>) {
+    private fun onPaymentMethodsResponseChanged(paymentMethods: List<PaymentMethod>, storedPaymentMethods: List<StoredPaymentMethod>) {
         Logger.d(TAG, "onPaymentMethodsResponseChanged")
+        setupStoredPaymentMethods(storedPaymentMethods)
+        setupPaymentMethods(paymentMethods)
+    }
+
+    private fun setupStoredPaymentMethods(storedPaymentMethods: List<StoredPaymentMethod>) {
+        storedPaymentMethodsList.clear()
+        for (storedPaymentMethod in storedPaymentMethods) {
+            val type = storedPaymentMethod.type
+            val id = storedPaymentMethod.id
+            if (type != null && id != null && PaymentMethodTypes.SUPPORTED_PAYMENT_METHODS.contains(type)) {
+                // We don't check for availability on stored payment methods
+                storedPaymentMethodsList.add(makeStoredModel(storedPaymentMethod))
+            } else {
+                Logger.e(TAG, "Unsupported stored payment method - $type - $id")
+            }
+        }
+    }
+
+    private fun makeStoredModel(storedPaymentMethod: StoredPaymentMethod): StoredPaymentMethodModel {
+        return when (storedPaymentMethod.type) {
+            PaymentMethodTypes.SCHEME -> {
+                StoredPaymentMethodModel(
+                    storedPaymentMethod.id ?: "",
+                    "•••• ${storedPaymentMethod.lastFour}",
+                    "${storedPaymentMethod.expiryMonth}/${DateUtils.removeFirstTwoDigitFromYear(storedPaymentMethod.expiryYear)}",
```

Comment:
Do we have a formatting util for this one? If not maybe we should extract one, or ideally let the UI do the formatting also. What do you think?

Reply:
Yep, let's do it.